app/src/org/commcare/dalvik/application/CommCareApplication.java

```diff
@@ -348,17 +357,55 @@ private void setRoots() {
     }
 
     private int initializeAppResources() {
-        //There should be exactly one of these for now
+        // Before we try to initialize a new app, check if any existing apps were left in a
+        // partially deleted state, and finish uninstalling them if so
         for (ApplicationRecord record : getGlobalStorage(ApplicationRecord.class)) {
-            if (record.getStatus() == ApplicationRecord.STATUS_INSTALLED) {
-                //We have an app record ready to go
+            if (record.getStatus() == ApplicationRecord.STATUS_DELETE_REQUESTED) {
+                record.uninstall(this);

```

Comment:
I wanna flag here that if this fails the app won't be able to start up, which doesn't necessarily seem good.

It is likely worth try/catching each of these uninstalls and logging failures. Having an app record around which can't uninstall is unfortunate but not a disaster, so we probably don't want to block app startup if it doesn't succeed.